### PR TITLE
Add methods for extracting true footprint for sampling valid data only

### LINKED ISSUE #1330: How to avoid nodata-only patches
### Summary

Large scenes are often rotated due to CRS like so:

Landsat scene

When using our GeoSamplers, we often sample from the nodata regions around the edges.

### Rationale

Sampling these nodata-only patches results in slower I/O and slower GPU training, despite these patches contributing nothing to the model.

### Implementation

There are two places where we could potentially improve performance.

### I/O

Best case scenario would be to avoid sampling these regions entirely. In #449, someone tried adding a check to the sampler that actually loads the patch, checks if it's entirely nodata pixels, and skips it if so. Unfortunately, this doesn't seem to work in parallel, and is slow since it needs to load each patch twice.

If there was a way to load the image in its native CRS (i.e., a square with no nodata pixels in the orientation it was taken in), this would solve all of our problems. I don't know of a way to do this.

### GPU

This is actually easier to solve. We could add a feature to our data module base classes that removes all nodata-only images from each mini-batch inside `transfer_batch_to_device` or another step. This would result in variable batch sizes, but I don't think that's an issue.

### Alternatives

_No response_

### Additional information

This is a highly requested feature:

* #449 
* #548 
* #1170

### PR DESCRIPTION
**Pull request by adriantre**
Fix #1330 

RasterDatasets may contain nodata regions due to projecting all file to the same CRS, and due to eventual inherit nodata regions in the images.
When IntersectionDataset joins this with VectorDataset, this may yield 
1. false positive samples (bad for learning)
2. empty negative samples (may be bad for learning)

The solution can be summarised as:
- [ ] In RasterDataset, when opening each file, extract footprint and add to rtree index object
- [ ] In IntersectionDataset._merge_dataset_indices copy over the footprint to the new index. 
- [ ] In the same method, could optimise by minimizing bbox to cover only actual intersection of valid data.
- [ ] In RandomGeoSampler.__iter__, use this footprint to validate that sample bbox actually overlaps, and don't yield until a valid box is found. 
- [ ] Enable the same for GridGeoSampler **(probably other PR)**
- [ ] Remove label mask for eventual nodata-regions that  outside regions in boundary. (As the criteria above is `overlaps` and not `contains`, corners of the resulting sample may still contain nodata, while the label mask still may cover this.) **(probably other PR)**
- [ ] Add ability to balance positive and negative samples. The VectorData can be intersected with the raster valid data footprint in the GeoSampler to facilitate ballancing positives and negatives. Right now torchgeo gives the user no control of this. **(probably other PR)**

Useful resources:
Rasterio nodata masks:
https://rasterio.readthedocs.io/en/latest/topics/masks.html#nodata-masks

Extract valid data footprint as vector
https://gist.github.com/sgillies/9713809

Reproject valid data footprint with rasterio
https://geopandas.org/en/stable/docs/user_guide/reproject_fiona.html#rasterio-example

### CODE REVIEW

File: torchgeo/datasets/utils.py [hunk 1]
```diff
@@ -847,3 +851,20 @@ def array_to_tensor(array: np.typing.NDArray[Any]) -> Tensor:
     elif array.dtype == np.uint32:
         array = array.astype(np.int64)
     return torch.tensor(array)
+
+
+def valid_data_footprint_from_datasource(
+    src: rasterio.io.DatasetReader, destination_crs: pyproj.crs.crs.CRS
+) -> Polygon:
+    # Read valid/nodata-mask
+    mask = src.read_masks()
```
Comment (johnnv1) [inline torchgeo/datasets/utils.py:860]: `read_masks()` returns `C, H, W` shape it has more than one band/mask

probably needs to do something like

```python
if len(masks) == 3:
    mask = np.logical_or(*masks[:, ...])
```

File: torchgeo/datasets/utils.py [hunk 2]
```diff
@@ -847,3 +851,20 @@ def array_to_tensor(array: np.typing.NDArray[Any]) -> Tensor:
...
+    # Close holes
+    sieved_mask = sieve(mask, 500)
```
Comment (johnnv1) [inline torchgeo/datasets/utils.py:862]: we didn't have to know the minimum `size` accepted for any raster... Maybe it can have a `factor` to compute this based on the mask shape?

File: torchgeo/datasets/utils.py [hunk 3]
```diff
@@ -847,3 +851,20 @@ def array_to_tensor(array: np.typing.NDArray[Any]) -> Tensor:
...
+    # extract polygon for valid data values
+    geom = next(g for g, v in shapes(sieved_mask, transform=src.transform) if v > 0)
```
Comment (johnnv1) [inline torchgeo/datasets/utils.py:864]: instead of the first polygon, maybe we should go for the bigger polygon? Or if we aim to handle everything, maybe it should be a join between the geometries

File: torchgeo/samplers/single.py [hunk 4]
```diff
@@ -142,10 +142,9 @@ def __iter__(self) -> Iterator[BoundingBox]:
             hit = self.hits[idx]
             bounds = BoundingBox(*hit.bounds)
 
-            # Choose a random index within that tile
-            bounding_box = get_random_bounding_box(bounds, self.size, self.res)
-
-            yield bounding_box
+            yield get_random_bounding_box_check_valid_overlap(
```
Comment (johnnv1) [inline torchgeo/samplers/single.py:145]: maybe first add it as an option in the API, and pass the max_retries too

File: torchgeo/datasets/utils.py (hunk 1 above)
Comment (adriantre) [reply torchgeo/datasets/utils.py:860]: We should choose to only use the mask for one spectral band, and it should probably work well, like `mask=src.read_masks(1)`. But in case there is a spatial shift between the bands, passing all of them to `sieve` will (to my understanding) find the valid mask across **all** bands.

Comment (adriantre) [reply torchgeo/datasets/utils.py:860]: So I think your proposal is covered by `sieve`?

Comment (adamjstewart) [reply torchgeo/datasets/utils.py:860]: 1 spectral band may not be enough, especially in the case of Landsat 7 SLC-off. Would be curious to know how far things shift. If it's only a few pixels, then that's within the margin of error of this method.

File: torchgeo/datasets/utils.py (hunk 2 above)
Comment (adriantre) [reply torchgeo/datasets/utils.py:862]: The target here is a polygon with no holes. Probably 500 is never too big (22 x 22 pixels). Could increase it, too.

If there are more (bigger) holes left, we could close them using shapely after converting to vector. What do you think?

File: torchgeo/datasets/utils.py (hunk 1 above)
Comment (adriantre) [reply torchgeo/datasets/utils.py:860]: Looking at this again I see that we indeed need the logical_or to cover shifts between spectral bands. Will push soon!

(My last comment about `sieve` (or `shape`) solving this was not correct)

File: torchgeo/datasets/utils.py (hunk 3 above)
Comment (adriantre) [reply torchgeo/datasets/utils.py:864]: I am now pushing a suggestion where we keep all polygons (that are not nodata) and make a multipolygon. This will then support "complex" footprints, or disjunct mosaics

File: torchgeo/samplers/single.py (hunk 4 above)
Comment (adriantre) [reply torchgeo/samplers/single.py]: Pushing a suggestion now

File: torchgeo/datasets/geo.py [hunk 5]
```diff
@@ -569,11 +579,21 @@ def _load_warp_file(self, filepath: str) -> DatasetReader:
...
         if src.crs != self.crs:
-            vrt = WarpedVRT(src, crs=self.crs)
+            valid_nodatavals = [val for val in src.nodatavals if val is not None]
+            if not valid_nodatavals:  # Case for Sentinel2 L1C
+                nodata = 0.0  # Is it safe to assume? For Sentinel2 L1C it is 0.0
+            else:
+                # Get the first valid nodata value.
+                # Usualy the same value for all bands
+                nodata = valid_nodatavals[0]
+            vrt = WarpedVRT(src, nodata=nodata, crs=self.crs)
```
Comment (adriantre) [inline torchgeo/datasets/geo.py:596]: Sentinel-2 has as far I can see no value set for nodata. I looked everywhere. Even enabling [alpha-layer](https://gdal.org/drivers/raster/sentinel2.html#open-options) and opening the [MSK_QUALIT-file] in the Sentinel-2 gdal driver, and looking through the (https://sentinels.copernicus.eu/web/sentinel/technical-guides/sentinel-2-msi/level-1c/masks) I found nothing.

Comment (adriantre) [reply torchgeo/datasets/geo.py:596]: This change will set the nodata-value. Some datasets have other nodata-values, and we should probably let the user overwrite this, for example in their RasterDataset.

Comment (adriantre) [reply torchgeo/datasets/geo.py:596]: Currently, the nodata is only overridden for the warped datasources. The non-warped (already correct CRS) are opened as is, but would also need to have the nodata overridden.

File: torchgeo/samplers/single.py [hunk 6]
```diff
@@ -142,10 +156,16 @@ def __iter__(self) -> Iterator[BoundingBox]:
             hit = self.hits[idx]
             bounds = BoundingBox(*hit.bounds)
```
Comment (adriantre) [inline torchgeo/samplers/single.py:157]: Only the first hit (file) is chosen. Currently I use this file to extract the footprint. But the sample is read from the merged raster, and the footprint for this one file might not cover the other.

Should be fairly easy to combine the footprints for all hits and send that to `get_random_bounding_box_check_valid_overlap`. 

Let me know if my assumptions are correct, and I will implement it.

File: torchgeo/datasets/utils.py (hunk 3 above)
Comment (johnnv1) [reply torchgeo/datasets/utils.py]: > I am now pushing a suggestion where we keep all polygons (that are not nodata) and make a multipolygon. This will then support "complex" footprints, or disjunct mosaics

Nice, I would give it as a suggestion because I see some cases like this on a daily basis 😄

File: torchgeo/datasets/utils.py (hunk 2 above)
Comment (adriantre) [reply torchgeo/datasets/utils.py:867]: Hmm, you are probably on to something. I'm struggling to decide what effect it might have if we set size too big or too small

File: torchgeo/samplers/single.py (hunk 6 above)
Comment (adriantre) [reply torchgeo/samplers/single.py:157]: The footprint for a hit is static. So this could be joined and cropped during init of IntersectionDataset

Comment (adriantre) [reply torchgeo/samplers/single.py:157]: I assume I have understood this correctly, and added this functionality in `_merge_dataset_indices`.

File: torchgeo/datasets/utils.py (hunk 2 above)
Comment (johnnv1) [reply torchgeo/datasets/utils.py:867]: I thought of something like:  closing holes that are bigger than the window size, we can still be getting some cases retrieving samples with just nodata ... considering the multi-polygon thing here.

One example is if we beforehand masks Sentinel-2 clouds as no data when closing the holes considering a size bigger than the window size we still can get random samples inside/within this nodata regions

Comment (adriantre) [reply torchgeo/datasets/utils.py:867]: Sounds smart!

Comment (adriantre) [reply torchgeo/datasets/utils.py:867]: One thing is that the desired patch_size to be used by sampler is not available at this point in the code. This happens on RasterDataset init, separate from the Sampler init.

**Comment by adamjstewart** [comment]
I'm trying to get back up to speed on this. Can you rebase? Also, would be good to test with the new I/O Bench dataset I added. See https://torchgeo.readthedocs.io/en/latest/user/contributing.html#i-o-benchmarking for instructions.

**Comment by adriantre** [comment]
> I'm trying to get back up to speed on this. Can you rebase? Also, would be good to test with the new I/O Bench dataset I added. See https://torchgeo.readthedocs.io/en/latest/user/contributing.html#i-o-benchmarking for instructions.

I rebased now, have a look. I'll look into your new dataset!

**Comment by adriantre** [comment]
I did a quick test.

My initial implementation is slower on your benchmark dataset `io_raw`. 
I assume it is for two reasons, as `setup` and `dataloader_next` is the bad apples:
1. Computing the true_footprints takes about 2 seconds. (This can be precomputed and stored/cached)
2. My `RandomGeoSampler` is naive, as it just tries until it finds a random box that overlaps with true_footprint, which may result in thousands of re-tries. 

Having `patch_size=128` and `num_workers = 0`:
6 seconds slower in total.
Reduces the number of no-data only samples from 1206 to 0.

What threw me off a little was that it seems like a lot of the nodata pixels for some reason get value`1.`, is this correct?
This can e.g. be observed in `RasterDataset.__getitem__` under `if self.is_image`.

GPU available: True (mps), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs

File: torchgeo/datasets/utils.py (hunk 1 above)
Comment (adriantre) [reply torchgeo/datasets/utils.py:869]: `np.logical_or.reduce(mask, axis=0).astype('uint8')` now accept any number of bands.

**Comment by adamjstewart** [comment]
So if I'm understanding correctly, the current state of this PR is that the naïve implementation of this is actually slower than loading all nodata pixels? In that case, we may want to push this to the 0.7.0 milestone so we have more time to ensure that everything is bug-free.

May be worth looking into how Raster Vision does this, they've been around longer than us: https://github.com/microsoft/torchgeo/issues/1330#issuecomment-2284444522

P.S. Need to resolve merge conflicts.

**Comment by adriantre** [comment]
If the only goal is low speed per epoch, then yes. But if the goal is low speed for the model to converge, this might still be faster.

In my opinion, this PR is the first step towards #1883 which is really important for faster overall training.

**Comment by adamjstewart** [comment]
I guess I would need to see additional tests to verify if it is actually faster for model convergence. Unfortunately that's even harder to time since it's inherently stochastic. If we can't also make it faster to load data, it seems unlikely to be merged.

**Comment by adriantre** [comment]
> There's no mechanism for on-the-fly filtering based on nodata content in RV GeoDatasets. However, RV also provides a way to pre-chip a dataset and then treat it as a non-geospatial dataset. And that does allow filtering based on nodata percentage via a [nodata_threshold option](https://docs.rastervision.io/en/stable/api_reference/_generated/rastervision.core.rv_pipeline.chip_options.ChipOptions.html#rastervision.core.rv_pipeline.chip_options.ChipOptions.nodata_threshold).

*from https://github.com/microsoft/torchgeo/issues/1330#issuecomment-2284444522*

This "pre-chipping" is my idea for how this could be solved, but my idea still keeps it as a GeoDataset. 

> Alternatives
One good alternative/evolution that I see is replacing the rtree-index created in IntersectionDataset. Instead of the current rtree-index, add the vectorized raster-footprints and/or vector data (actual features/shapes from e.g. shapefiles) to GeoPandas GeoDataFrames, and use their implementation of rtree to rapidly find areas where they intersect. The temporal dimension would not be part of the index (not supported), but can be added as a filter once spatial matches are found. At the end of [this article](https://geoffboeing.com/2016/10/r-tree-spatial-index-python/) they split their polygons [raster-footprints] into small grid cells [of patch size] and can rapidly retrieve cells [samples] where there is overlap with the other data set [labels], or no overlap. The underlying rtree-indices in RasterDatasets and VectorDatasets will be the same, and still be used to read the data upon sampling.

*from first post in #1883*

I would need an indication that this is a good/promising solution before starting to dive into it.

**Comment by adamjstewart** [comment]
Both of these would be pretty big core changes and would need a lot of justification. Losing explicit temporal sampling would be unfortunate, as we're trying to improve time series support in the next 6 months. But the ability to check for overlap of arbitrary polygons (not just boxes) would be very nice. I am worried about the decrease in performance though, especially as you approach millions of images. A formal mathematical R-tree will always be significantly faster than a hacky R-tree that supports arbitrary shapes. Basically, I don't know the best way to do this, and I don't want you to sink tens of hours into something that may never be accepted, but I also wish someone would solve this problem and don't have the time/need myself.

**Comment by adamjstewart** [comment]
Note that we are considering porting from RTree to Geopandas or STAC, which support complex shapely Geometries and would allow us to make use of this true footprint. May be worth revisiting this after the port is finished.

**Comment by adamjstewart** [comment]
#2747 is now merged if you want to revive this PR. We can now store arbitrary Polygons directly in our GeoDataFrame. Sampling will come later.

**Comment by adamjstewart** [comment]
Superseded by #2903
